From 0bcda0a77b2999e660e67aa658ffad426ab668be Mon Sep 17 00:00:00 2001
From: Jaroslav Rohel <jrohel@redhat.com>
Date: Tue, 18 Jun 2019 09:40:03 +0200
Subject: [PATCH 1/2] [config] Fix: toString() to not insert [] (RhBug:1584442)

OptionStringList::toString() returned a string starting with
"[" and ended with "]" that represents list of strings (Python
compatibility like print([<values>])). But opposite method
OptionStringList::fromString() assumes input without "[" and "]".
The dnf.conf and .repo files do not support value surrounded
by "[" and "]" therefore this commit removes these characters.

Fix https://bugzilla.redhat.com/show_bug.cgi?id=1584442
---
 libdnf/conf/OptionStringList.cpp         | 2 --
 libdnf/module/ModulePackageContainer.cpp | 1 -
 2 files changed, 3 deletions(-)

diff --git a/libdnf/conf/OptionStringList.cpp b/libdnf/conf/OptionStringList.cpp
index 9fbc79a1..7f618e4e 100644
--- a/libdnf/conf/OptionStringList.cpp
+++ b/libdnf/conf/OptionStringList.cpp
@@ -97,7 +97,6 @@ void OptionStringList::set(Priority priority, const std::string & value)
 std::string OptionStringList::toString(const ValueType & value) const
 {
     std::ostringstream oss;
-    oss << "[";
     bool next{false};
     for (auto & val : value) {
         if (next)
@@ -106,7 +105,6 @@ std::string OptionStringList::toString(const ValueType & value) const
             next = true;
         oss << val;
     }
-    oss << "]";
     return oss.str();
 }
 
diff --git a/libdnf/module/ModulePackageContainer.cpp b/libdnf/module/ModulePackageContainer.cpp
index 1105576e..0ee32039 100644
--- a/libdnf/module/ModulePackageContainer.cpp
+++ b/libdnf/module/ModulePackageContainer.cpp
@@ -1187,7 +1187,6 @@ bool ModulePackageContainer::Impl::ModulePersistor::update(const std::string & n
 
     OptionStringList slist{std::vector<std::string>()};
     auto profiles = slist.toString(getProfiles(name));
-    profiles = profiles.substr(1, profiles.size()-2);
     if (!parser.hasOption(name, "profiles") || parser.getValue(name, "profiles") != profiles) {
         parser.setValue(name, "profiles", std::move(profiles));
         changed = true;
-- 
2.21.0


From b7bf705a03011c9f610fbe413f7c57647ee1ab1e Mon Sep 17 00:00:00 2001
From: Jaroslav Rohel <jrohel@redhat.com>
Date: Tue, 18 Jun 2019 13:42:27 +0200
Subject: [PATCH 2/2] [module] update() compare parsed value instead of raw
 string

---
 libdnf/module/ModulePackageContainer.cpp | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/libdnf/module/ModulePackageContainer.cpp b/libdnf/module/ModulePackageContainer.cpp
index 0ee32039..d6c86e55 100644
--- a/libdnf/module/ModulePackageContainer.cpp
+++ b/libdnf/module/ModulePackageContainer.cpp
@@ -1185,10 +1185,10 @@ bool ModulePackageContainer::Impl::ModulePersistor::update(const std::string & n
         changed = true;
     }
 
-    OptionStringList slist{std::vector<std::string>()};
-    auto profiles = slist.toString(getProfiles(name));
-    if (!parser.hasOption(name, "profiles") || parser.getValue(name, "profiles") != profiles) {
-        parser.setValue(name, "profiles", std::move(profiles));
+    OptionStringList profiles{getProfiles(name)};
+    if (!parser.hasOption(name, "profiles") ||
+        OptionStringList(parser.getValue(name, "profiles")).getValue() != profiles.getValue()) {
+        parser.setValue(name, "profiles", profiles.getValueString());
         changed = true;
     }
 
-- 
2.21.0

